File: dart/lib/src/utils/get_pyth_feed_account_key.dart
import 'dart:typed_data';
import 'package:solana/solana.dart' hide RpcClient;
import '../constants/addresses.dart';

/// Gets the Pyth feed account key
///
/// This function mirrors js/src/utils/getPythFeedAccountKey.ts
///
/// [shard] - The shard number
/// [priceFeed] - The price feed as list of integers
///
/// Returns the derived Pyth feed account public key
Future<Ed25519HDPublicKey> getPythFeedAccountKey(
  int shard,
  List<int> priceFeed,
) async {
  final buffer = ByteData(2);
  buffer.setUint16(0, shard, Endian.little);

  final seeds = [
    buffer.buffer.asUint8List(),
    Uint8List.fromList(priceFeed),
  ];

  final programAddress = Ed25519HDPublicKey.fromBase58(defaultPythPushProgram);

  final result = await Ed25519HDPublicKey.findProgramAddress(
    seeds: seeds,
    programId: programAddress,
  );

  return result;
}

File: js/src/utils/getPythFeedAccountKey.ts
import { Buffer } from "buffer";
import { PublicKey } from "@solana/web3.js";
import { DEFAULT_PYTH_PUSH_PROGRAM } from "../constants";

/**
 * Derives a Pyth push-oracle price-feed account address from its shard and feed ID.
 *
 * @param shard Pyth feed shard identifier
 * @param priceFeed Price-feed identifier bytes
 * @returns Derived Pyth feed address and bump seed
 *
 * @example
 * ```ts
 * const [address] = getPythFeedAccountKey(0, priceFeedId);
 * ```
 */
export const getPythFeedAccountKey = (shard: number, priceFeed: number[]) => {
  const buffer = Buffer.alloc(2);
  buffer.writeUint16LE(shard);
  return PublicKey.findProgramAddressSync(
    [buffer, Buffer.from(priceFeed)],
    DEFAULT_PYTH_PUSH_PROGRAM,
  );
};
